### PR TITLE
Add a11y Storybook addon

### PR DESCRIPTION
**Pull request by Tom-Bonnike**
Hi! I watched Pedro’s presentation of Modulz this afternoon and it looks amazing, good job! It prompted me to play around with radix and do some quick accessibility checks to test the primitives 🙈 

# This PR

Although I’m guessing you’ll move away from Storybook at some point, I’ve noticed a few accessibility issues in the Radix components. This PR sets up a [storybook addon](https://www.npmjs.com/package/@storybook/addon-a11y) that should help avoid those in the future.

The addon uses [axe](https://github.com/dequelabs/axe-core) under the hood, which is an accessibility testing engine. It can sometimes report false positives but I didn’t notice any while playing around, and you can totally customize axe to ignore some errors, so IMO there is never a strong reason not to install this addon, but that’s obviously up to you! I set it up during my tests so I figured I could just open a PR.

Here’s an example of what it looks like when browsing the Button component:
Screenshot 2020-04-09 at 18 20 45

It also adds a color blindness control which can be handy:
Screenshot 2020-04-09 at 18 39 06

Will Modulz include something similar, or at least a color contrast check? That’d be a great feature. :)

# Accessibility issues

I know Radix and your primitives are WIP but here’s a non exhaustive list of a11y issues I found, in no particular order (some of them are not reported by the addon).

## In Radix

- The `blue` variant doesn’t pass the WCAG contrast ratio test (in both `Button` and `Badge`). Same for the `normal` variant of `Code`, the `highlight` variant of `Link`, the `fade` variant of the `ToggleButton` and the `active` variant of `List`.
- The waiting button has a focus outline but it’s almost invisible.
- The `Switch` and `Checkbox` components don’t have focus states, but you have some opened issues for those already. There’s no issue for the `Slider` component, though, which also lacks a focus state.
- Your icons are not accessible since they do not label the `svg`s correctly. But that’s probably an issue for the `radix-icons` repository?

Happy to open separate issues for those.

## Directly in the primitives

- The inputs/textareas are not using (or enforcing the use of) `label` elements. You should probably add a visually hidden label to those controls by default. The `placeholder` attribute is [not](https://www.smashingmagazine.com/2018/06/placeholder-attribute/#interoperability) [good](https://www.nngroup.com/articles/form-design-placeholders/) [enough](https://www.deque.com/blog/accessible-forms-the-problem-with-placeholders/). If you decide to keep using `placeholder`, you could start by improving its color contrast so that it passes the WCAG contrast ratio check.
- The checkboxes, sliders and switches don’t use `label` elements either (or don’t link their `p` label using `aria-labelledby`), which actually leads to UX issues like the fact that clicking the label of the checkboxes doesn’t toggle the checkbox.
- The `disabled` attribute is used for buttons/inputs but it leads to them being undiscoverable since those are not focusable, which could lead to screen-reader users getting confused. 😞The solution here would be to output `aria-disabled="true"` instead, style them yourself and make sure all mouse and keyboard events are ignored.

I’m guessing the same primitives are also used in Modulz projects? Do you think accessibility should be handled in user-land, which I guess also gives users more flexibility on the DOM structure? IMO if it could be enforced directly through the primitives it’d be awesome, but you probably have given more thought to this than I did. 😅
I also would have loved to contribute to the primitives, hope you’ll make them open-source at some point!

### CODE REVIEW

File: packages/radix/.storybook/config.js [hunk 1]
```diff
@@ -12,6 +13,7 @@ function ThemeComponent({ children }) {
 }
 
 addDecorator((story, context) => <ThemeComponent>{story()}</ThemeComponent>);
+addDecorator(withA11y);
```
Comment (Tom-Bonnike) [inline packages/radix/.storybook/config.js:16]: This adds the a11y decorator to all stories.

File: packages/radix/package.json [hunk 2]
```diff
@@ -41,6 +41,7 @@
   "devDependencies": {
     "@babel/core": "^7.7.5",
     "@modulz/radix-icons": "2.0.0",
+    "@storybook/addon-a11y": "5.2.8",
     "@storybook/addon-options": "5.2.8",
```
Comment (Tom-Bonnike) [inline packages/radix/package.json:45]: It seems like you’re not using this one, btw?

**Comment by benoitgrelard** [comment]
Hey @Tom-Bonnike Thanks for this! The storybook addon looks cool, might add it to our lower-level primitives repo.

I'll respond to a few of the things you brought up down here:

# Radix

> The blue variant doesn’t pass the WCAG contrast ratio test (in both Button and Badge). Same for the normal variant of Code, the highlight variant of Link, the fade variant of the ToggleButton and the active variant of List.

> The waiting button has a focus outline but it’s almost invisible.

Regarding color contrast accessibility, I'll bring in @vladmoroz our designer to the discussion.

> The Switch and Checkbox components don’t have focus states, but you have some opened issues for those already. There’s no issue for the Slider component, though, which also lacks a focus state.

Yep, 100% right, we have open issues to look at these. I'll add one for the `Slider` now, but it might be because this one is due to be re-implemented in a custom way. At the moment it uses `<input type="range />` which makes it impractical to style some states/parts.

> Your icons are not accessible since they do not label the svgs correctly. But that’s probably an issue for the radix-icons repository?

Yes by default these are just SVG's really. I believe the accessibility for these should come at a higher level, once we implement something like an `IconButton` that's where it'll be baked in most likely with an `aria-label` and and `aria-hidden` on the SVG. Was that what you had in mind?

# Primitives

> The inputs/textareas are not using (or enforcing the use of) label elements. You should probably add a visually hidden label to those controls by default. The placeholder attribute is not good enough. If you decide to keep using placeholder, you could start by improving its color contrast so that it passes the WCAG contrast ratio check.

> The checkboxes, sliders and switches don’t use label elements either (or don’t link their p label using aria-labelledby), which actually leads to UX issues like the fact that clicking the label of the checkboxes doesn’t toggle the checkbox.

100% agree for using real labels rather than placeholder! 💯
The reason nothing in baked into those is that we currently see these as the real low-level leaf components that you can style. So they are literally just the control (input, checkbox, radio, etc).
I think the accessibility would come rather from a higher-level component like a `Field` where you can compose a label and a control, etc.

> The disabled attribute is used for buttons/inputs but it leads to them being undiscoverable since those are not focusable, which could lead to screen-reader users getting confused. 😞The solution here would be to output aria-disabled="true" instead, style them yourself and make sure all mouse and keyboard events are ignored.

That's a great shout. We are actually doing some of this in places already like `Tabs` so screen readers are aware of disabled tabs. We'll give some thoughts and considerations to extending this to other controls. In your experience, are there any downsides in doing this for most controls like button, input, checkbox, etc?

Thanks again for you time @Tom-Bonnike!

**Comment by Tom-Bonnike** [comment]
> The reason nothing in baked into those is that we currently see these as the real low-level leaf components that you can style. So they are literally just the control (input, checkbox, radio, etc).
I think the accessibility would come rather from a higher-level component like a Field where you can compose a label and a control, etc.

I see, makes total sense, but do you think you’ll have higher-level primitives for those in Modulz, or do you see accessibility as your user’s responsibility?
That’s fair, but I’d personally love to see a design tool that actually enforces accessibility or at least forces people to think about it directly at the design stage, at least for “simpler” stuff like color contrast, forms and images. It would also also force designers to think about all states (e.g. errored fields, which also require a bit of work to be accessible). So yes, this PR is me lobbying for Modulz to make my job as a developer easier 👹Not sure how that would all look like in practice without limiting the users too much, though.

> Yes by default these are just SVG's really. I believe the accessibility for these should come at a higher level, once we implement something like an IconButton that's where it'll be baked in most likely with an aria-label and and aria-hidden on the SVG. Was that what you had in mind?

Here’s the simplified `Icon` component we used at my previous job, we had an accessibility audit done recently which confirmed we were doing things correctly:
```js
// Assuming TextAlignLeftIcon is a component rendering a <svg> tag, we used it like this:
// <Icon icon={TextAlignLeftIcon} /> 

const Icon = props => {
  if (!props.icon) return

  if (!props.label) {
    return (
      <props.icon
        aria-hidden
        focusable='false'
      />
    )
  }

  // The `role` and `aria-live` attributes might be useful in case the icon
  // needs to provide further information to the accessibility tree such as
  // for a loading spinner.
  return (
    <span
      title={props.label}
      className={props.className}
      role={props.role}
      aria-live={props['aria-live']}
    >
      <props.icon
        aria-hidden
        focusable='false'
      />
      <VisuallyHidden>{props.label}</VisuallyHidden>
    </span>
  )
}
```

Although you don’t need to label decorative icons when they’re accompanying text, apparently adding `aria-hidden` and `focusable="false"` on svgs is recommended to have them properly ignored by all screen-readers.
But, for example, the first example with the icon in your ToggleButton story is not accessible, obviously it’s “just” a Storybook story and not a real use-case, but if the icons had a required `label` prop, it would be impossible to forget about it, and it still could be manually set to an empty string when you’re sure you don’t want one.

> We are actually doing some of this in places already like Tabs so screen readers are aware of disabled tabs.

Great 👍 

> In your experience, are there any downsides in doing this for most controls like button, input, checkbox, etc?

I was surprised to discover this after an accessibility audit at my last job, but because using `disabled` on buttons and inputs make them unfocusable, it can lead screen-reader users to dead end because 
they won’t easily be able to discover them and so won’t be able to submit the form if there isn’t any hint. 
Disabled inputs are imo anyway an antipattern, unless having an unactionable yet visible dimmed field is what you actually want, but in most cases you’re probably looking for a `readonly` input instead 😄 
Form buttons should always be actionable and make users jump back to the field that’s invalid, or there should be something that explains clearly why the button is disabled.
We had “fake” disabled buttons with `aria-disabled="true"` that were styled like a disabled one and ignored clicks (`pointer-events: none`) and keypresses on them for places where we _really_ needed a disabled button.

**Comment by benoitgrelard** [comment]
❤️ this @Tom-Bonnike, keep it coming!

> I see, makes total sense, but do you think you’ll have higher-level primitives for those in Modulz, or do you see accessibility as your user’s responsibility?
That would be fair, but I’d personally love to see a design tool that actually enforces accessibility or at least forces people to think about it directly at the design stage, at least for “simpler” stuff like color contrast, forms and images. It would also also force designers to think about all states (e.g. errored fields, which also require a bit of work to be accessible). So yes, this PR is me lobbying for Modulz to make my job as a developer easier 👹Not sure how that would all look like in practice without limiting the users too much, though.

I think some accessibility will always lie within the responsibility of the app developers because there are many things that can't be automated. That being said, it is most definitely our goal to automate and give for free as much as possible. So we would definitely be looking at these kinds of higher level components which wouldn't necessarily feature in Modulz itself but would be available in the code export.

An existing example of this already is our `DropdownMenu` component. It doesn't appear in Modulz, it's simplify a functional aid. What you see and get to style in Modulz is a `Menu` component (item, separator, etc). This `Menu` component gets used in a few places, like `DropdownMenu`, `RightClickMenu` (another example of functional aid) and under the hood of the `Select` component. The `DropdownMenu` then takes any button (could be one from your DS or not) and a `Menu` component and wires it all together for you with the proper accessibility (menu-button aria pattern). Feel free to take a look at it in the radix styleguide ([`DropdownMenu`](https://radix.modulz.app/docs/dropdown-menu/) component.).

> Yup! Here’s the simplified Icon component we used at my previous job, we had an accessibility audit done recently which confirmed we were doing things correctly:

I like this a lot. We currently don't have this because we only have radix icons available for us. The slightly longer term plan is to enable different types of icon libraries through Modulz and enabling users to pick an icon in places (think Tick for a checkbox, etc). I don't see why we can't start thinking about such a component though, love it. Love the extra attention for role, etc. I'm gonna put an issue for us to look at this, seems like a fairly easy win.

> Although you don’t need to label decorative icons when they’re accompanying text, apparently adding aria-hidden and focusable="false" on svgs is recommended to have them properly ignored by all screen-readers.

Yep, `aria-hidden` makes sense.
Can I ask about the `focusable='false'`? Is this a regular DOM attribute? A special prop?

> But, for example, the first example with the icon in your ToggleButton story is not accessible, obviously it’s “just” a Storybook story and not a real use-case, but if the icons had a required label prop, it would be impossible to forget about it, and it still could be manually set to an empty string when you’re sure you don’t want one.

Yep for sure stories won't always be nice and up to date.
Also some components don't appear in the Radix styleguide but we do have a `VisuallyHidden` component in the export and in that story we have:

```jsx
<>
  <p>Screen readers will enunciate the button below as "Notifications, button."</p>
  <button type="button">
    <VisuallyHidden>Notifications</VisuallyHidden>
    <BoxIcon aria-hidden />
  </button>
</>
```

> We had “fake” disabled buttons with aria-disabled="true" that were styled like a disabled one and ignored clicks (pointer-events: none) and keypresses on them for places where we really needed a disabled button.

Yeah that's interesting, that's partly what we do in `Tabs` for example. And interestingly, we've also recently updated most of these controls primitive to use `pointer-events: none` when disabled for styling reason (because it simplifies things like making sure hover styles can't work on disabled controls, I've never understood why the browser lets that happen.)

👍

**Comment by Tom-Bonnike** [comment]
> That being said, it is most definitely our goal to automate and give for free as much as possible.

Good to hear 👍

> Can I ask about the focusable='false'? Is this a regular DOM attribute? A special prop?

I looked up online because I couldn’t remember and it’s an attribute that’s only valid on `<svg>`s. Apparently some browsers consider them as focusable elements by default (IE11 and Edge I think?) and this helps make sure they can never receive focus.

Cool, sounds good, I hope to get the chance to use Modulz in a team soon. :)

**Comment by vladmoroz** [comment]
> The blue variant doesn’t pass the WCAG contrast ratio test (in both Button and Badge). Same for the normal variant of Code, the highlight variant of Link, the fade variant of the ToggleButton and the active variant of List.

Thanks for the catch @Tom-Bonnike, I think we only tested this blue color on white only, and not on the off-white that these components use. We are taking a look at colors soon, will make sure to update it.

> The waiting button has a focus outline but it’s almost invisible.

Didn't even know! That slipped through. We are plan to review these buttons at some point and replace the waiting variant with a more suitable component.

Let me know if you saw anything else, thank you 🙏

**Pull request closed without merging**